src/storage-client/src/controller/remap_migration.rs

```diff
@@ -0,0 +1,358 @@
+use tokio::sync::Mutex;
+use tracing::info;
+
+use mz_expr::PartitionId;
+use mz_persist_client::cache::PersistClientCache;
+use mz_persist_client::read::ReadHandle;
+use mz_persist_client::{PersistClient, ShardId, Upper};
+use mz_repr::{Datum, Diff, GlobalId, TimestampManipulation};
+use mz_timely_util::order::Partitioned;
+
+use crate::types::sources::data::SourceData;
+use crate::types::sources::MzOffset;
+use crate::util::antichain::MutableOffsetAntichain;
+use crate::util::remap_handle::RemapHandle;
+
+use super::CollectionMetadata;
+
+/// A struct capable of performing remap shard migrations. See comment on
+/// [`RemapHandleMigrator::migrate`].
+pub struct RemapHandleMigrator<
+    T: Timestamp + Lattice + TotalOrder + Codec64 + From<EpochMillis> + TimestampManipulation,
+> {
+    persist_client: PersistClient,
+    current_shard_read_handle: ReadHandle<SourceData, (), T, Diff>,
+    as_of: Antichain<T>,
+    write_shard: ShardId,
+    partitioned_source: bool,
+}
+
+impl<T: Timestamp + Lattice + TotalOrder + Codec64 + From<EpochMillis> + TimestampManipulation>
+    RemapHandleMigrator<T>
+{
+    /// Prior to this commit, remap shards contained data of schema (Int32?,
+    /// Uint64). Int32 was present only for Kafka sources where it represented
+    /// the Kafka partition.
+    ///
+    /// This schema introduced an impedance mismatch between the remap data and
+    /// native Timely types; tl;dr is that an empty remap shard, if taken
+    /// directly to Timely time types, produces an empty Antichain (i.e. the
+    /// shard is closed from further modification), when in reality we wanted
+    /// it to express the minimum Antichain (i.e. any future input is valid).
+    ///
+    /// TO resolve this, we designed a migration to instead use the schema
+    /// (Range<Int32>?, Uint64). Range<Int32> still represents Kafka
+    /// partitions, but instead is meant to represent seen Kafka partitions, as
+    /// well as the range of those still unseen, e.g.:
+    ///
+    /// (Bottom, 0) -> 0 // trivially empty sub-zero elements in domain
+    /// [0, 1) -> 12 // partition 0 offset
+    /// [1, Top] -> 0 // all unseen partitions
+    ///
+    /// In addition to migrating Kafka sources to this schema, we also
+    /// introduce the mimum element (i.e. the 0 lsn) for Postgres sources that
+    /// are not new but whose remap shards are empty.
+    pub async fn migrate(
+        persist_clients: Arc<Mutex<PersistClientCache>>,
+        read_metadata: CollectionMetadata,
+        write_shard: ShardId,
+        // additional information to improve logging
+        id: GlobalId,
+        partitioned_source: bool,
+    ) -> Option<()> {
+        let mut migrator = Self::new(
+            persist_clients,
+            read_metadata.clone(),
+            write_shard,
+            id,
+            partitioned_source,
+        )
+        .await
+        .unwrap()?;
+
+        let (migrated_form, upper) = migrator.next().await?;
```

Comment:
related to the `as_of` comment. Since we start at `since` and we must continue to process until `upper` this should be in a loop where we read a migrated batch and then compare and append it to the new shard until we reach the current upper

Reply:
It's a little cleaner to treat this set of updates like a multi-part snapshot, essentially performing the migration in memory first. From the code perspective, this means putting the loop inside the `next` method rather than outside of the `next`, `compare_and_append` pair.

Do you think that's all right, or is the batching preferable?

Reply:
Or maybe the "full fidelity" needs to include all of the distinct times of the updates and if we skip ahead, we're essentially forcing some kind of consolidation-like behavior to occur, which we're trying to avoid. Will work on writing this with the outer for loop as you suggested.

Reply:
Sorry for the thrashing on these comments. It turns out that the non-partitioned migration requires that we read the entire shard's content before determining if we should insert `MzOffset::from(0)` at the original `since`. This means we need to do the whole migration in memory first and then write it to the shard all at once.

Reply:
Ah, I was also trying to avoid writing to the new remap shard unless we knew we were migrating. Your approach makes sense, though, given that leaking shards is not a huge deal.